File: RubyCivitas/lib/vista_textual.rb
#encoding:utf-8
require 'io/console'
require_relative "civitas_juego"
require_relative "controlador"

module Civitas

  class Vista_textual

    def initialize(civitas)
      @juego_model = civitas
      @i_gestion = -1
      @i_propiedad = -1
    end

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end

    def menu(titulo, lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end

    
    def comprar

      lista = [TipoRespuestas::NO,TipoRespuestas::SI]
      eleccion=menu("¿Desea comprar la calle a la que ha llegado?",["No","Si"])
      return lista[eleccion]

    end

    def gestionar

      lista_s=["Vender","Hipotecar","Cancelar hipoteca","Construir casa","Construir hotel","Terminar"]
      lista_g=[TiposGestionesInmobiliarias::VENDER,TiposGestionesInmobiliarias::HIPOTECAR,TiposGestionesInmobiliarias::CANCELAR_HIPOTECA,TiposGestionesInmobiliarias::CONSTRUIR_CASA,TiposGestionesInmobiliarias::CONSTRUIR_HOTEL,TiposGestionesInmobiliarias::TERMINAR]

      seleccion= menu("¿Qué gestión desea realizar?",lista_s)
      @i_gestion=seleccion

      if @i_gestion != 5 && !@juego_model.get_jugador_actual.propiedades.empty?

        propiedades_nombre=[]
        propiedades_titulo=@juego_model.get_jugador_actual.propiedades

        for i in 0...propiedades_titulo.size() do
          propiedades_nombre.push(propiedades_titulo[i].nombre)
        end
        opcion = menu("Sobre qué propiedad?", propiedades_nombre)
        @i_propiedad=opcion
      end

    end

    def get_gestion
      @i_gestion
    end

    def get_propiedad
      @i_propiedad
    end

    def mostrar_siguiente_operacion(operacion)

      puts operacion.to_s

    end

    def mostrar_eventos

      while(Diario.instance.eventos_pendientes)

        puts Diario.instance.leer_evento

      end
    end

    def set_civitas_juego(civitas)
      @juego_model = civitas
      #self.actualizar_vista
    end

    def actualizar_vista
      
      puts @juego_model.get_jugador_actual.to_string
      puts @juego_model.get_casilla_actual.to_string

    end

    def salir_carcel
      titulo = "Elige la forma para intentar salir de la carcel"
      lista = [Salidas_carcel::PAGANDO, Salidas_carcel::TIRANDO]
      opcion = menu(titulo,["Pagando", "Tirando el dado"])
      return lista[opcion]
    end

    attr_reader :i_gestion, :i_propiedad
    public :mostrar_eventos, :gestionar

    
  end

end

File: civitasjava/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.Respuestas;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;


class VistaTextual {
	CivitasJuego juegoModel;
	int iGestion = -1;
	int iPropiedad = -1;
	private static String separador = "==================";
	
	private Scanner in;

	public VistaTextual(){
		in = new Scanner(System.in);
	}
	
	void mostrarEstado(String estado){
		System.out.println(estado);
	}
	
	void pausa(){
		System.out.println("Pulsa una tecla: ");
		in.nextLine();
	}
	
	int leeEntero(int max, String msg1, String msg2){
		boolean hecho;
		String cad;
		int num=-1;
		do{
			System.out.println(msg1);
			cad = in.nextLine();
			try{
				num=Integer.parseInt(cad);
				hecho = true;
			}
			catch(NumberFormatException e){ //Entero no introducido
				System.out.println(msg2);
				hecho = false;
			}
			if(hecho && (num<0 || num>=max)){
				System.out.println(msg2);
				hecho = false;
			}			
		}while(!hecho);
		
		return num;	
	}
	
	
	int menu(String titulo, ArrayList<String> lista){
		String tab = "	";
		int opcion;
		System.out.println(titulo);
		for(int i=0; i<lista.size(); i++){
			System.out.println(tab + i + "-" + lista.get(i));
		}
		
		opcion = leeEntero(lista.size(), "\n" + tab + "Elige una opcion: ",
					tab + "Valor erroneo");
		return opcion;
	}
	
	
	SalidasCarcel salirCarcel(){
		int opcion = menu("Elige la forma para intentar salir de la carcel", 
				new ArrayList<> (Arrays.asList("Pagando", "Tirando el dado")));
		
		return (SalidasCarcel.values()[opcion]);
	}
	
	
	Respuestas comprar(){
		int opcion = menu ("¿Quiere comprar la casilla en la que ha caido?",
				new ArrayList<>(Arrays.asList("SI", "NO")));
		
		return(Respuestas.values()[opcion]);
	}
	
	
	void gestionar(){
		int opcion =  menu("¿Que numero de gestion inmobiliaria quiere realizar?",
                        new ArrayList<>(Arrays.asList("Vender", "Hipotecar", 
                                "Cancelar hipoteca", "Construir casa",
                                "Construir hotel", "Terminar")));
		iGestion = opcion;
		
		ArrayList<String> calles = new ArrayList<String>();
		
		for(int i=0; i<juegoModel.getJugadorActual().getPropiedades().size(); i++){
			calles.add(juegoModel.getJugadorActual().
                                getPropiedades().get(i).getNombre());
		}
		
		iPropiedad = menu("¿Sobre que casilla quiere realizar la"
                        + " operacion?", calles);
	}
	
	public int getGestion(){
		return iGestion;
	}
	
	public int getPropiedad(){
		return iPropiedad;
	}
	
	
	void mostrarSiguienteOperacion(OperacionesJuego operacion){
		if(operacion == OperacionesJuego.AVANZAR){
			System.out.println("La siguiente operacion es avanzar");
		}
		else if(operacion == OperacionesJuego.COMPRAR){
			System.out.println("La siguiente operacion es comprar");
		}
		else if(operacion == OperacionesJuego.GESTIONAR){
			System.out.println("La siguiente operacion es gestionar");
		}
		else if(operacion == OperacionesJuego.PASAR_TURNO){
			System.out.println("La siguiente operacion es pasar turno");
		}
		else{
			System.out.println("La siguiente operacion es salir de la carcel");
		}
	}
	
	
	
	void mostrarEventos(){
		while(Diario.getInstance().eventosPendientes()){
			System.out.println(Diario.getInstance().leerEvento());
		}	
	}
	
	public void setCivitasJuego(CivitasJuego civitas){
		juegoModel = civitas;
	}
	
	void actualizarVista(){
            System.out.println(juegoModel.infoJugadorTexto());
        }
	
}